cmd/kubeadm/app/cmd/reset_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+)
+
+func TestNewResetData(t *testing.T) {
+	testCases := []struct {
+		name        string
+		args        []string
+		flags       map[string]string
+		validate    func(*testing.T, *resetData)
+		expectError bool
+		data        *resetData
+	}{
+		{
+			name: "flags parsed correctly",
+			flags: map[string]string{
+				options.CertificatesDir:       "/tmp",
+				options.NodeCRISocket:         "unix:///var/run/crio/crio.sock",
+				options.IgnorePreflightErrors: "all",
+				options.ForceReset:            "true",
+				options.DryRun:                "true",
+				options.CleanupTmpDir:         "true",
+			},
+			data: &resetData{
+				certificatesDir:       "/tmp",
+				criSocketPath:         "unix:///var/run/crio/crio.sock",
+				ignorePreflightErrors: sets.NewString("all"),
+				forceReset:            true,
+				dryRun:                true,
+				cleanupTmpDir:         true,
+			},
+		},
+		{
+			name: "fails if preflight ignores all but has individual check",
+			flags: map[string]string{
+				options.IgnorePreflightErrors: "all,something-else",
+			},
+			expectError: true,
+		},
+		{
+			name: "pre-flights errors from CLI args",
+			flags: map[string]string{
+				options.IgnorePreflightErrors: "a,b",
+			},
+			validate: expectedResetIgnorePreflightErrors(sets.NewString("a", "b")),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// initialize an external reset option and inject it to the reset cmd
+			resetOptions := newResetOptions()
+			cmd := newCmdReset(nil, nil, resetOptions)
+
+			// sets cmd flags (that will be reflected on the reset options)
+			for f, v := range tc.flags {
+				cmd.Flags().Set(f, v)
+			}
+
+			// test newResetData method
+			data, err := newResetData(cmd, resetOptions, nil, nil)
+			if err != nil && !tc.expectError {
+				t.Fatalf("newresetData returned unexpected error: %v", err)
+			}
+			if err == nil && tc.expectError {
+				t.Fatalf("newresetData didn't return error when expected")
```

Comment:
newResetData
same problem above

Reply:
this can be

```
if (err != nil) != expected ...

// expected err got ...
```

Reply:
hmm... I see your point, but we can compare the error here directly, or else we need assemble exactly the same error which is make the test here looks ugly and complicated.

Reply:
But I updated the testcase to validate the error string to make sure the error is what we want.

Reply:
> make the test here looks ugly and complicated.

it's normal, and we have it in a number of places

Reply:
> if (err != nil) != expected

this is just a pseudocode, right? this won't compile since we cannot check the equal between bool with `error` or string.

> have it in a number of places

could you pls give me an example, so I can see is that applicable for this one as well.

BTW, I have update the `expectError` to be a string, so I can check the error.msg(), it's not an `error` type anymore. Because if I want to compare two `error` I need to build two `error` with the same error string, one of them is here:
https://github.com/kubernetes/kubernetes/blob/76ccb69cc1d9aeae4e5f808424d1cf2d411eb708/cmd/kubeadm/app/apis/kubeadm/validation/validation.go#L618

Reply:
https://github.com/kubernetes/kubernetes/blob/673ff9730a90224a3a4b0449e68073df5ebf78bc/cmd/kubeadm/app/apis/bootstraptoken/v1/utils_test.go#L76-L77

Reply:
okay, that field is defined as `bool`, this is not.

Reply:
```
			if (err != nil) != (len(rt. expectedError) != 0) {
```

Reply:
the benefit for it is being a `error` or a error string is that if there are multiple failure points with different failure reasons, we can validate that this does fail with the expected failure reason.

I can switch it to `bool` if you think that is enough.

Reply:
```golang
  if (err != nil) != (len(tc.expectError) != 0) {
	 if err != nil && !strings.Contains(err.Error(), tc.expectError) {
		t.Fatalf("newResetData returned unexpected error, expected: %s, got %v", tc.expectError, err)
	 }
	t.Fatalf("newResetData didn't return error expected %s", tc.expectError)
     }
```

@pacoxu , did you mean something like the code above? seems like the code is not simplified a lot.

Reply:
> this can be
> 
> ```
> if (err != nil) != expected ...
> 
> // expected err got ...
> ```

Following @neolit123 's comments, the current change is ok. Leave it to Lubomir to double-check if the current change is like what he thought. 😄
LGTM